docs/development-guide.md

```diff
@@ -0,0 +1,123 @@
+and how to work with the code base.
+
+Before you dive into this, it is best to read:
+
+* The whole [Usage Guide][usage-guide]
+* The [Code of Conduct][code of conduct]
+* The [Contributing][contributing] guide
+
+## Docker
+
+The Columbo project uses Docker to ease setting up a consistent development environment. The Docker documentation has
+details on how to [install docker][install-docker] on your computer.
+
+Once that is configured, the test suite can be run locally:
+
+```bash
+docker-compose run --rm test
+```
+
+If you want to be able to execute code in the container:
+
+```bash
+docker-compose run --rm devbox
+(your code here)
+```
+
+In the devbox environment you'll be able to enter a python shell and import `columbo` or any dependencies.
+
+## Debugging
+
+The docker container has [pdb++][pdbpp-home] install that can be used as a debugger. (However, you are welcome to set up
+a different debugger if you would like.)
+
+This allows you to easily create a breakpoint anywhere in the code.
+
+```python
+def my_function():
+    breakpoint()
+    ...
+```
+
+When your the code, you will drop into an interactive `pdb++` debugger.
+
+See the documentation on [pdb][pdb-docs] and [pdb++][pdbpp-docs] for more information.
+
+## Testing
+
+You'll be unable to merge code unless the linting and tests pass. You can run these in your container via:
+
+```bash
+docker-compose run --rm test
+```
+
+This will run the same tests, linting, and code coverage that are run by the CI pipeline. The only difference is that,
+when run locally, `black` and `isort` are configured to automatically correct issues they detect.
+
+Generally we should endeavor to write tests for every feature. Every new feature branch should increase the test
+coverage rather than decreasing it.
+
+We use [pytest][pytest-docs] as our testing framework.
+
+#### Stages
+
+To customize / override a specific testing stage, please read the documentation specific to that tool:
+
+1. [PyTest][pytest-docs]
+2. [MyPy][mypy-docs]
+3. [Black][black-docs]
+4. [Isort][isort-docs]
+4. [Flake8][flake8-docs]
+5. [Bandit][bandit-docs]
+
+### `setup.py`
+
+Setuptools is used to packaging the library.
+
+**`setup.py` must not import anything from the package** When installing from source, the user may not have the
+packages dependencies installed, and importing the package is likely to raise an `ImportError`. For this reason, the
+**package version should be obtained without importing**. This is explains why `setup.py` uses a regular expression to
+grabs the version from `__init__.py` without actually importing.
+
+### Requirements
+
+* **requirements.txt** - Lists all direct dependencies (packages imported by the library).
+* **Requirements-test.txt** - Lists all direct requirements needed to run the test suite & lints.
+
+## Publishing the Package
+
+TODO: Not currently implemented.
```

Comment:
Out of scope for this PR, but since this is marked as TODO, wanted to plug a library like [`bump2version`](https://github.com/c4urself/bump2version/#installation), which might help add structure to the process of tagging & releasing new versions.

Reply:
Yeah. I know there are a few tools that can help with version bump management. I'll look into those as we get closer to 1.0.